plutus-scb/src/Plutus/SCB/Query.hs

```diff
@@ -128,3 +162,208 @@ utxoIndexProjection =
+    AwaitTxConfirmedRequest txids -> Just txids
+    _ -> Nothing
+
+-- | Open endpoints by contract instance
+userEndpointRequests :: forall t key position. Projection (IteratedContractState ActiveEndpoints) (StreamEvent key position (ChainEvent t))
+userEndpointRequests = iteratedContractStateProjection $ \case
+    UserEndpointRequest t -> Just t
+    _ -> Nothing
+
+-- | Contract instances' requests for "own" public keys
+ownPubKeyRequests :: forall t key position. Projection (IteratedContractState ()) (StreamEvent key position (ChainEvent t))
+ownPubKeyRequests = iteratedContractStateProjection $ \case
+    OwnPubkeyRequest WaitingForPubKey -> Just ()
+    _ -> Nothing
+
+-- | Requests for subsets of the UTXO set
+utxoAtRequests :: forall t key position. Projection (IteratedContractState AddressSet) (StreamEvent key position (ChainEvent t))
+utxoAtRequests = iteratedContractStateProjection $ \case
+    UtxoAtRequest addr -> Just addr
+    _ -> Nothing
+
+-- | Requests to learn about the next transaction at an address
+nextTxAtRequests :: forall t key position. Projection (IteratedContractState AddressSet) (StreamEvent key position (ChainEvent t))
+nextTxAtRequests = iteratedContractStateProjection $ \case
+    NextTxAtRequest addr -> Just addr
+    _ -> Nothing
+
+-- | Requests to balance, sign and submit unbalanced transactions
+writeTxRequests :: forall t key position. Projection (IteratedContractState PendingTransactions) (StreamEvent key position (ChainEvent t))
+writeTxRequests = iteratedContractStateProjection $ \case
+    WriteTxRequest pendingTxs -> Just pendingTxs
+    _ -> Nothing
+
+-- | The 'ContractResponse' messages that a contract has received in the current iteration.
+data ContractInboxState =
+    ContractInboxState
+        { cisIterations    :: ContractIterationState
+        , cisInboxMessages :: Map ContractInstanceId (Seq ContractResponse)
+        }
+
+-- | Messages sent to contract instances
+inboxMessages :: forall t key position. Projection ContractInboxState (StreamEvent key position (ChainEvent t))
+inboxMessages =
+    let projectionEventHandler s@ContractInboxState{cisIterations, cisInboxMessages} = \case
+            (StreamEvent _ _ (ContractEvent (ContractInstanceStateUpdateEvent e))) ->
+                ContractInboxState
+                    { cisIterations = updateContractIterationState e cisIterations
+                    , cisInboxMessages = Map.delete (csContract e) cisInboxMessages
+                    }
+            (StreamEvent _ _ (ContractEvent (ContractInboxEvent (ContractInboxMessage{ciInstance,ciIteration,ciPayload}))))
+                | Just (Max ciIteration) == Map.lookup ciInstance (unContractIterationState cisIterations) ->
+                    ContractInboxState
+                        { cisIterations
+                        , cisInboxMessages = Map.insertWith (<>) ciInstance (Seq.singleton ciPayload) cisInboxMessages
+                        }
+            _ -> s
```

Comment:
This looks suspiciously similar to 235-245. Is there something to factor out here?

Reply:
I merged them by shuffling the data types for the contract events around a little and then using prisms for accessing the value that we are interested in